File: solutions/typescript/072-smallest-range-covering-elements-from-k-lists.ts
/**
 * Problem 72: Smallest Range Covering Elements from K Lists (LeetCode 632)
 * Difficulty: Hard
 * Language: TypeScript
 */
function smallestRange(nums: number[][]): number[] {
  const tagged: [number, number][] = nums.flatMap((list, i) => list.map(v => [v, i] as [number, number]));
  const sorted = [...tagged].sort((a, b) => a[0] - b[0]);
  const fold = sorted.reduce<{ best: [number, number]; counts: Map<number, number>; covered: number; left: number }>(
    (acc, [val, grp], right) => {
      const c = new Map(acc.counts);
      const prev = c.get(grp) ?? 0;
      c.set(grp, prev + 1);
      let covered = acc.covered + (prev === 0 ? 1 : 0);
      let left = acc.left;
      let best = acc.best;
      const shrink = (l: number, cov: number, cs: Map<number, number>, b: [number, number]): { left: number; covered: number; counts: Map<number, number>; best: [number, number] } => {
        if (cov < nums.length) return { left: l, covered: cov, counts: cs, best: b };
        const newB: [number, number] = (val - sorted[l][0] < b[1] - b[0]) ? [sorted[l][0], val] : b;
        const lg = sorted[l][1];
        const nc = new Map(cs);
        nc.set(lg, nc.get(lg)! - 1);
        const newCov = nc.get(lg)! === 0 ? cov - 1 : cov;
        return shrink(l + 1, newCov, nc, newB);
      };
      const res = shrink(left, covered, c, best);
      return { best: res.best, counts: res.counts, covered: res.covered, left: res.left };
    },
    { best: [-Infinity, Infinity], counts: new Map(), covered: 0, left: 0 }
  );
  return fold.best;
}

File: solutions/rust/072-smallest-range-covering-elements-from-k-lists.rs
/**
 * Problem 72: Smallest Range Covering Elements from K Lists (LeetCode 632)
 * Difficulty: Hard
 * Language: Rust
 */
fn smallest_range(nums: Vec<Vec<i32>>) -> Vec<i32> {
    let k = nums.len();
    let mut tagged: Vec<(i32, usize)> = nums.iter().enumerate()
        .flat_map(|(i, list)| list.iter().map(move |&v| (v, i))).collect();
    tagged.sort();
    let (mut left, mut best) = (0, (i32::MIN, i32::MAX));
    let mut counts = vec![0usize; k];
    let mut covered = 0usize;
    for &(v, g) in &tagged {
        if counts[g] == 0 { covered += 1; }
        counts[g] += 1;
        while covered == k {
            let (lv, lg) = tagged[left];
            if v - lv < best.1 - best.0 { best = (lv, v); }
            counts[lg] -= 1;
            if counts[lg] == 0 { covered -= 1; }
            left += 1;
        }
    }
    vec![best.0, best.1]
}
